rfcs/0116-artifact-service.md

```diff
@@ -0,0 +1,196 @@
+# RFC 116 Artifact Service
+* Comments: [#116](https://api.github.com/repos/taskcluster/taskcluster-rfcs/issues/116)
+* Proposed by: @jhford
+
+# Summary
+
+Taskcluster tasks often product Artifacts during the course of being run.
+Examples include browser archives, json metadata and log files.  We currently
+store Artifacts in a single region of S3 and split management of Artifacts
+between the Queue and Cloud Mirror.  This project is to take Artifact handling
+out of the Queue and into its own service that takes over the responsibilities
+for cross-region and cross-cloud replication.
+
+The Artifact service should not be directly exposed to end users as a Rest API.
+It will instead be a generalised API that the Queue uses to facilitate artifact
+creation from tasks.  In other words, the Queue gates operations to the
+Artifact service.  The service will be exposed to endusers by means of a `Location`
+header value on a `300` series redirect.
+
+The Queue will now store only the name of Artifacts and would treat the name of
+an Artifact as an opaque identifier and defer all management of the actual
+bytes to this system.  The Queue would also stop doing IP to Object store
+mappings.
+
+This RFC is to decide on what the Artifacts service is and its interface, not
+implementation details
+
+## Motivation
+
+* We want the Queue to be easier to understand, develop and maintain
+* We want Artifacts to be easier to understand, develop and maintain
+* We want better Artifact handling without increasing complexity of the Queue
+  codebase
+* We want to have more clear responsibilities for our services
+* We want to consolidate the handling of Artifacts into a single service
+* We want to build generalised software when possible as participants of a free
+  software community
+
+# Details
+
+* The Artifact service will start from the Queue's `artifacts.js` file
+* Auth might not be done with scopes and tc-auth, but a shared secret between
+  Queue and Artifacts
+* The Queue will block new tasks after task resolution by not issuing new
+  signed urls
+* Data storage in Artifacts will be done with Postgres and not Azure Table
+  Storage
+* We do not intend to have an operation to list artifacts in the API
+* The redirecting logic of the Artifact service will adhere to the HTTP spec
+  using 300-series redirects where possible
+* Where content-negotiation is impossible, the service will block creation of
+  artifacts when incoming requests aren't compatible with required response
```

Comment:
respond `406` if resource is gzipped and `Accept: gzip` is missing...

Reply:
I still don't understand it - respond 406 on which request?

Reply:
so you're Client A and requesting resource X which is stored with gzip content-encoding in s3.  You submit a request with the headers `{}` (i.e. none).  Because we know that we're going to serve a response which *must* have `{Content-Encoding: gzip}`, but we don't know for sure if you can accept gzip encoding, we fail with and send a 406.  When the client is updated to send headers `{Accept-Encoding: gzip}`, we send the 300 series redirect to S3.

The request which would possibly send the 406 would be the first from this service.  The queue would 302 to this service, the service would look up the underying storage and see that it's going to force a content encoding but that necoding isn't in the accept-encoding header, then this service would send a 406.

Reply:
Thanks, that makes a lot more sense.  Probably including a scenario like that in the doc would help :)

Reply:
Will do!